// @fluentui/react-aria/lib/activedescendant/useActiveDescendant.js
import * as React from 'react';
import { useOptionWalker } from './useOptionWalker';
import { ACTIVEDESCENDANT_ATTRIBUTE } from './constants';
export function useActiveDescendant(options) {
    const { imperativeRef, matchOption } = options;
    const activeParentRef = React.useRef(null);
    const { listboxRef, optionWalker } = useOptionWalker({
        matchOption
    });
    const getActiveDescendant = ()=>{
        var _listboxRef_current;
        return (_listboxRef_current = listboxRef.current) === null || _listboxRef_current === void 0 ? void 0 : _listboxRef_current.querySelector(`[${ACTIVEDESCENDANT_ATTRIBUTE}]`);
    };
    const scrollActiveIntoView = (active)=>{
        if (!listboxRef.current) {
            return;
        }
        if (listboxRef.current.offsetHeight >= listboxRef.current.scrollHeight) {
            return;
        }
        const { offsetHeight, offsetTop } = active;
        const { offsetHeight: parentOffsetHeight, scrollTop } = listboxRef.current;
        const isAbove = offsetTop < scrollTop;
        const isBelow = offsetTop + offsetHeight > scrollTop + parentOffsetHeight;
        const buffer = 2;
        if (isAbove) {
            listboxRef.current.scrollTo(0, offsetTop - buffer);
        }
        if (isBelow) {
            listboxRef.current.scrollTo(0, offsetTop - parentOffsetHeight + offsetHeight + buffer);
        }
    };
    const setActiveDescendant = (nextActive)=>{
        const active = getActiveDescendant();
        if (active) {
            active.removeAttribute(ACTIVEDESCENDANT_ATTRIBUTE);
        }
        if (nextActive) {
            var _activeParentRef_current;
            nextActive.setAttribute(ACTIVEDESCENDANT_ATTRIBUTE, '');
            scrollActiveIntoView(nextActive);
            (_activeParentRef_current = activeParentRef.current) === null || _activeParentRef_current === void 0 ? void 0 : _activeParentRef_current.setAttribute('aria-activedescendant', nextActive.id);
        } else {
            var _activeParentRef_current1;
            (_activeParentRef_current1 = activeParentRef.current) === null || _activeParentRef_current1 === void 0 ? void 0 : _activeParentRef_current1.removeAttribute('aria-activedescendant');
        }
    };
    React.useImperativeHandle(imperativeRef, ()=>({
            first: ()=>{
                if (!listboxRef.current || !activeParentRef.current) {
                    return;
                }
                optionWalker.setCurrent(listboxRef.current);
                const first = optionWalker.first();
                if (first) {
                    setActiveDescendant(first);
                }
            },
            next: ()=>{
                if (!listboxRef.current || !activeParentRef.current) {
                    return;
                }
                const active = getActiveDescendant();
                if (!active) {
                    return;
                }
                optionWalker.setCurrent(active);
                const next = optionWalker.next();
                if (next) {
                    setActiveDescendant(next);
                }
            },
            prev: ()=>{
                if (!listboxRef.current || !activeParentRef.current) {
                    return;
                }
                const active = getActiveDescendant();
                if (!active) {
                    return;
                }
                optionWalker.setCurrent(active);
                if (!matchOption(active)) {
                    optionWalker.prev();
                }
                const next = optionWalker.prev();
                if (next && next !== listboxRef.current) {
                    setActiveDescendant(next);
                }
            },
            blur: ()=>{
                if (!activeParentRef.current) {
                    return;
                }
                setActiveDescendant(undefined);
            },
            active: ()=>{
                if (listboxRef.current) {
                    var _getActiveDescendant;
                    return (_getActiveDescendant = getActiveDescendant()) === null || _getActiveDescendant === void 0 ? void 0 : _getActiveDescendant.id;
                }
            },
            focus: (id)=>{
                if (!listboxRef.current) {
                    return;
                }
                optionWalker.setCurrent(listboxRef.current);
                let cur = optionWalker.next();
                while(cur && cur.id !== id){
                    cur = optionWalker.next();
                }
                if (cur) {
                    setActiveDescendant(cur);
                }
            }
        }));
    return {
        listboxRef,
        activeParentRef
    };
}
